packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol

```diff
@@ -241,6 +257,38 @@ contract FaultDisputeGame is IFaultDisputeGame, Clone, Semver {
         move(_parentIndex, _claim, false);
     }
 
+    /// @inheritdoc IFaultDisputeGame
+    function addLocalData(uint256 _ident, uint256 _partOffset) external {
+        // INVARIANT: Local data can only be added if the game is currently in progress.
+        if (status != GameStatus.IN_PROGRESS) revert GameNotInProgress();
+
+        IPreimageOracle oracle = VM.oracle();
+        if (_ident == 1) {
+            // Load the L1 head hash into the game's local context in the preimage oracle.
+            oracle.loadLocalData(_ident, Hash.unwrap(l1Head), 32, _partOffset);
+        } else if (_ident == 2) {
+            // Load the earliest output root that commits to the passed L2 block number
+            // into the game's local context in the preimage oracle.
+            Types.OutputProposal memory proposal = L2_OUTPUT_ORACLE.getL2OutputAfter(
+                l2BlockNumber()
+            );
```

Comment:
Shouldn't this be the last output prior to the l2 block number rather than the first after it? Note that this is the agreed l2 output root to start execution from. I'm a little unclear exactly what the l2 block number being passed in is representing here though - is it what's being challenged or the agreed starting point?

Ultimately we want cannon to run from the output root immediately prior to the one being challenged. If you disagree with the immediately prior root you should just challenge it instead of the later one as they'll then both be removed. I wonder if it's easier to just pass in the output root index being challenged from the oracle and then we can load the claim being challenged, the block number its from and the prior claim all from the output oracle.

Reply:
Ah, `getL2OutputAfter` is after or equal to the block number.  So if the passed in block number is the agreed starting point, then this would work.  The claimed output root should then be from the output immediately after this.

It would be more future proof to specify the output being challenged by index and so that when outputs are permissionless it can handle having multiple conflicting outputs for the same block number and get the "parent" of the claim as the starting point since there may be multiple competing children of an output.

Reply:
ah, the catch with using the output index is that if output roots are deleted the indices may be reused. That would be ok if the deletion checks the output root matches but not ok if we are only loading the data on demand in either the game contract or the challenger since the index may have changed in the meantime.

Using block number has the same issue though since `getL2OutputAfter` could return a different output root to what it did when the root claim was calculated.

Reply:
Hm this is interesting, should probably be addressed in another PR. The `L2OutputOracle` probably needs a rework that will allow for this - great points. Ideally we do away with the array and move to a mapping or other data structure. This could future proof it for permissionless output proposals as well.

Alternatively we can add the deletion check if we don't want to dig into the L2OO for the alpha - this was the current plan. Prior to deleting the output root, we'd check if the output root at the given index has changed, and if so, the game would be invalidated. We'd need to persist the output root on game creation rather than lazily fetching it in this function, but that's no problem.